File: tests/benchmarks/closure_benchmark.lua
#!/usr/bin/env lua

-- Comprehensive Orus Language Benchmark - Lua Version
-- Tests all currently supported features in one complex benchmark

print("=== Orus Comprehensive Performance Benchmark ===")

local start_time = math.floor(os.clock() * 1000000) -- microseconds

-- === ARITHMETIC OPERATIONS INTENSIVE TEST ===
print("Phase 1: Intensive Arithmetic Operations")

-- Complex arithmetic chains
local a = 100
local b = 50
local c = 25
local d = 12
local e = 6

-- Multi-step calculations
local result1 = a + b * c - math.floor(d / e)
local result2 = math.floor((a - b) * (c + d) / e)
local result3 = math.floor(a * b / c) + d - e
local result4 = math.floor((a + b + c + d + e) / (a - b - c - d - e))

-- Nested arithmetic expressions
local complex1 = math.floor(((a + b) * c - d) / ((e + d) * c - b))
local complex2 = math.floor((a * b + c * d) / (a - b + c - d))
local complex3 = math.floor((math.floor(a / b) * c + d) - (math.floor(e * d / c) + b))

-- Iterative calculations (simulating loops)
local sum_val = 0
local counter = 0
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter
counter = counter + 1
sum_val = sum_val + counter

print("Arithmetic Phase Results:")
print("Complex result 1:", complex1)
print("Complex result 2:", complex2)
print("Complex result 3:", complex3)
print("Iterative sum:", sum_val)

-- === VARIABLE OPERATIONS INTENSIVE TEST ===
print("Phase 2: Intensive Variable Operations")

-- Variable assignment chains
local var1 = 10
local var2 = var1 * 2
local var3 = var2 + var1
local var4 = var3 - var2
local var5 = var4 * var1
local var6 = math.floor(var5 / var2)
local var7 = var6 + var3
local var8 = var7 - var4
local var9 = var8 * var5
local var10 = math.floor(var9 / var6)

-- Variable swapping simulation
local temp_a = var1
local temp_b = var2
var1 = temp_b
var2 = temp_a

-- Complex variable interdependencies
local base = 5
local multiplier = 3
local offset = 7
local threshold = 20

local calc_a = base * multiplier + offset
local calc_b = calc_a - threshold
local calc_c = math.floor(calc_b * base / multiplier)
local calc_d = calc_c + offset - threshold
local calc_e = math.floor(calc_d * multiplier / base)

-- Variable reuse patterns
local accumulator = 0
accumulator = accumulator + base
accumulator = accumulator * multiplier
accumulator = accumulator - offset
accumulator = accumulator + threshold
accumulator = math.floor(accumulator / base)
accumulator = accumulator * offset
accumulator = accumulator - multiplier

print("Variable Phase Results:")
print("Final var10:", var10)
print("Swapped var1:", var1)
print("Swapped var2:", var2)
print("Calc chain result:", calc_e)
print("Accumulator result:", accumulator)

-- === LITERAL VALUES INTENSIVE TEST ===
print("Phase 3: Intensive Literal Operations")

-- Large number calculations
local big1 = 1000000
local big2 = 500000
local big3 = 250000
local big4 = 125000
local big5 = 62500

-- Operations with large numbers
local large_sum = big1 + big2 + big3 + big4 + big5
local large_diff = big1 - big2 - big3 - big4 - big5
local large_prod = math.floor(big1 / 1000) * math.floor(big2 / 1000)
local large_div = math.floor(big1 / big2) * math.floor(big3 / big4)

-- Mixed literal and variable operations
local mixed1 = 42 + var1 * 17 - 8
local mixed2 = 99 - math.floor(var2 / 3) + 21
local mixed3 = 77 * var3 + 13 - 56
local mixed4 = math.floor(88 / var4) - 44 + 33

-- Decimal-like operations (using integer division)
local decimal1 = math.floor(100 * 355 / 113)
local decimal2 = math.floor(1000 * 22 / 7)
local decimal3 = math.floor(10000 * 618 / 1000)

print("Literal Phase Results:")
print("Large sum:", large_sum)
print("Large difference:", large_diff)
print("Large product:", large_prod)
print("Mixed calculation 1:", mixed1)
print("Mixed calculation 2:", mixed2)
print("Decimal approximation 1:", decimal1)
print("Decimal approximation 2:", decimal2)

-- === TIMESTAMP OPERATIONS INTENSIVE TEST ===
print("Phase 4: Intensive Timestamp Operations")

-- Multiple timestamp measurements
local ts1 = math.floor(os.clock() * 1000000)
local ts2 = math.floor(os.clock() * 1000000)
local ts3 = math.floor(os.clock() * 1000000)
local ts4 = math.floor(os.clock() * 1000000)
local ts5 = math.floor(os.clock() * 1000000)

-- Timestamp arithmetic
local ts_diff1 = ts2 - ts1
local ts_diff2 = ts3 - ts2
local ts_diff3 = ts4 - ts3
local ts_diff4 = ts5 - ts4

-- Complex timestamp calculations
local ts_total = ts_diff1 + ts_diff2 + ts_diff3 + ts_diff4
local ts_avg = math.floor(ts_total / 4)
local ts_max = ts_diff1
local temp_check = ts_diff2 - ts_max
temp_check = ts_diff3 - ts_max
temp_check = ts_diff4 - ts_max

-- Timestamp with arithmetic operations
local ts_calc1 = ts1 + 1000
local ts_calc2 = math.floor(ts2 * 2 / 2)
local ts_calc3 = ts3 - 500 + 500
local ts_calc4 = math.floor((ts4 + ts5) / 2)

print("Timestamp Phase Results:")
print("Timestamp 1:", ts1)
print("Timestamp 5:", ts5)
print("Total time diff:", ts_total)
print("Average time diff:", ts_avg)
print("Complex timestamp calc:", ts_calc4)

-- === PRINT OPERATIONS INTENSIVE TEST ===
print("Phase 5: Intensive Print Operations")

-- Multiple print statements with different data
print("Testing multiple print operations:")
print("Number:", 42)
print("Calculation:", 10 + 5)
print("Variable:", accumulator)
print("Expression:", (a + b) * c)
print("Large number:", big1)
print("Time:", ts1)

-- Print with complex expressions
print("Complex expression 1:", math.floor(((100 + 200) * 3) / 2))
print("Complex expression 2:", (500 - 100) * 2 + 50)
print("Complex expression 3:", math.floor(1000 / 10) + math.floor(200 / 4))

-- === COMPREHENSIVE INTEGRATION TEST ===
print("Phase 6: Comprehensive Integration")

-- Combine all features in complex calculations
local integration_start = math.floor(os.clock() * 1000000)

-- Complex integrated calculation
local step1 = math.floor(big1 / 1000) + (var1 * var2)
local step2 = step1 - (complex1 * 10)
local step3 = step2 + (ts_avg * 100)
local step4 = math.floor(step3 / (accumulator + 1))
local step5 = step4 * math.floor(decimal1 / 100)

-- Multiple variable interactions
local chain_a = step1 + step2
local chain_b = step3 + step4
local chain_c = step5 + chain_a
local chain_d = chain_b + chain_c
local final_result = math.floor(chain_d / 4)

-- Final timestamp
local integration_end = math.floor(os.clock() * 1000000)
local integration_time = integration_end - integration_start

print("Integration Phase Results:")
print("Step 1:", step1)
print("Step 2:", step2)
print("Step 3:", step3)
print("Step 4:", step4)
print("Step 5:", step5)
print("Final integrated result:", final_result)
print("Integration time:", integration_time)

-- === FINAL BENCHMARK RESULTS ===
local end_time = math.floor(os.clock() * 1000000)
local total_elapsed = end_time - start_time

print("=== BENCHMARK COMPLETE ===")
print("Total execution time:", total_elapsed)
print("Operations completed: 200+")
print("Variables used: 50+")
print("Arithmetic operations: 100+")
print("Print statements: 40+")
print("Timestamp operations: 20+")
print("Final benchmark score:", total_elapsed + final_result)
print("=== Orus Comprehensive Benchmark Complete ===")

File: tests/benchmarks/closure_benchmark.js
#!/usr/bin/env node

// Comprehensive Orus Language Benchmark - JavaScript Version
// Tests all currently supported features in one complex benchmark

console.log("=== Orus Comprehensive Performance Benchmark ===");

const start_time = Math.floor(Date.now() * 1000); // microseconds

// === ARITHMETIC OPERATIONS INTENSIVE TEST ===
console.log("Phase 1: Intensive Arithmetic Operations");

// Complex arithmetic chains
let a = 100;
let b = 50;
let c = 25;
let d = 12;
let e = 6;

// Multi-step calculations
let result1 = a + b * c - Math.floor(d / e);
let result2 = Math.floor((a - b) * (c + d) / e);
let result3 = Math.floor(a * b / c) + d - e;
let result4 = Math.floor((a + b + c + d + e) / (a - b - c - d - e));

// Nested arithmetic expressions
let complex1 = Math.floor(((a + b) * c - d) / ((e + d) * c - b));
let complex2 = Math.floor((a * b + c * d) / (a - b + c - d));
let complex3 = Math.floor((Math.floor(a / b) * c + d) - (Math.floor(e * d / c) + b));

// Iterative calculations (simulating loops)
let sum_val = 0;
let counter = 0;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;
counter = counter + 1;
sum_val = sum_val + counter;

console.log("Arithmetic Phase Results:");
console.log("Complex result 1:", complex1);
console.log("Complex result 2:", complex2);
console.log("Complex result 3:", complex3);
console.log("Iterative sum:", sum_val);

// === VARIABLE OPERATIONS INTENSIVE TEST ===
console.log("Phase 2: Intensive Variable Operations");

// Variable assignment chains
let var1 = 10;
let var2 = var1 * 2;
let var3 = var2 + var1;
let var4 = var3 - var2;
let var5 = var4 * var1;
let var6 = Math.floor(var5 / var2);
let var7 = var6 + var3;
let var8 = var7 - var4;
let var9 = var8 * var5;
let var10 = Math.floor(var9 / var6);

// Variable swapping simulation
let temp_a = var1;
let temp_b = var2;
var1 = temp_b;
var2 = temp_a;

// Complex variable interdependencies
let base = 5;
let multiplier = 3;
let offset = 7;
let threshold = 20;

let calc_a = base * multiplier + offset;
let calc_b = calc_a - threshold;
let calc_c = Math.floor(calc_b * base / multiplier);
let calc_d = calc_c + offset - threshold;
let calc_e = Math.floor(calc_d * multiplier / base);

// Variable reuse patterns
let accumulator = 0;
accumulator = accumulator + base;
accumulator = accumulator * multiplier;
accumulator = accumulator - offset;
accumulator = accumulator + threshold;
accumulator = Math.floor(accumulator / base);
accumulator = accumulator * offset;
accumulator = accumulator - multiplier;

console.log("Variable Phase Results:");
console.log("Final var10:", var10);
console.log("Swapped var1:", var1);
console.log("Swapped var2:", var2);
console.log("Calc chain result:", calc_e);
console.log("Accumulator result:", accumulator);

// === LITERAL VALUES INTENSIVE TEST ===
console.log("Phase 3: Intensive Literal Operations");

// Large number calculations
let big1 = 1000000;
let big2 = 500000;
let big3 = 250000;
let big4 = 125000;
let big5 = 62500;

// Operations with large numbers
let large_sum = big1 + big2 + big3 + big4 + big5;
let large_diff = big1 - big2 - big3 - big4 - big5;
let large_prod = Math.floor(big1 / 1000) * Math.floor(big2 / 1000);
let large_div = Math.floor(big1 / big2) * Math.floor(big3 / big4);

// Mixed literal and variable operations
let mixed1 = 42 + var1 * 17 - 8;
let mixed2 = 99 - Math.floor(var2 / 3) + 21;
let mixed3 = 77 * var3 + 13 - 56;
let mixed4 = Math.floor(88 / var4) - 44 + 33;

// Decimal-like operations (using integer division)
let decimal1 = Math.floor(100 * 355 / 113);
let decimal2 = Math.floor(1000 * 22 / 7);
let decimal3 = Math.floor(10000 * 618 / 1000);

console.log("Literal Phase Results:");
console.log("Large sum:", large_sum);
console.log("Large difference:", large_diff);
console.log("Large product:", large_prod);
console.log("Mixed calculation 1:", mixed1);
console.log("Mixed calculation 2:", mixed2);
console.log("Decimal approximation 1:", decimal1);
console.log("Decimal approximation 2:", decimal2);

// === TIMESTAMP OPERATIONS INTENSIVE TEST ===
console.log("Phase 4: Intensive Timestamp Operations");

// Multiple timestamp measurements
let ts1 = Math.floor(Date.now() * 1000);
let ts2 = Math.floor(Date.now() * 1000);
let ts3 = Math.floor(Date.now() * 1000);
let ts4 = Math.floor(Date.now() * 1000);
let ts5 = Math.floor(Date.now() * 1000);

// Timestamp arithmetic
let ts_diff1 = ts2 - ts1;
let ts_diff2 = ts3 - ts2;
let ts_diff3 = ts4 - ts3;
let ts_diff4 = ts5 - ts4;

// Complex timestamp calculations
let ts_total = ts_diff1 + ts_diff2 + ts_diff3 + ts_diff4;
let ts_avg = Math.floor(ts_total / 4);
let ts_max = ts_diff1;
let temp_check = ts_diff2 - ts_max;
temp_check = ts_diff3 - ts_max;
temp_check = ts_diff4 - ts_max;

// Timestamp with arithmetic operations
let ts_calc1 = ts1 + 1000;
let ts_calc2 = Math.floor(ts2 * 2 / 2);
let ts_calc3 = ts3 - 500 + 500;
let ts_calc4 = Math.floor((ts4 + ts5) / 2);

console.log("Timestamp Phase Results:");
console.log("Timestamp 1:", ts1);
console.log("Timestamp 5:", ts5);
console.log("Total time diff:", ts_total);
console.log("Average time diff:", ts_avg);
console.log("Complex timestamp calc:", ts_calc4);

// === PRINT OPERATIONS INTENSIVE TEST ===
console.log("Phase 5: Intensive Print Operations");

// Multiple print statements with different data
console.log("Testing multiple print operations:");
console.log("Number:", 42);
console.log("Calculation:", 10 + 5);
console.log("Variable:", accumulator);
console.log("Expression:", (a + b) * c);
console.log("Large number:", big1);
console.log("Time:", ts1);

// Print with complex expressions
console.log("Complex expression 1:", Math.floor(((100 + 200) * 3) / 2));
console.log("Complex expression 2:", (500 - 100) * 2 + 50);
console.log("Complex expression 3:", Math.floor(1000 / 10) + Math.floor(200 / 4));

// === COMPREHENSIVE INTEGRATION TEST ===
console.log("Phase 6: Comprehensive Integration");

// Combine all features in complex calculations
let integration_start = Math.floor(Date.now() * 1000);

// Complex integrated calculation
let step1 = Math.floor(big1 / 1000) + (var1 * var2);
let step2 = step1 - (complex1 * 10);
let step3 = step2 + (ts_avg * 100);
let step4 = Math.floor(step3 / (accumulator + 1));
let step5 = step4 * Math.floor(decimal1 / 100);

// Multiple variable interactions
let chain_a = step1 + step2;
let chain_b = step3 + step4;
let chain_c = step5 + chain_a;
let chain_d = chain_b + chain_c;
let final_result = Math.floor(chain_d / 4);

// Final timestamp
let integration_end = Math.floor(Date.now() * 1000);
let integration_time = integration_end - integration_start;

console.log("Integration Phase Results:");
console.log("Step 1:", step1);
console.log("Step 2:", step2);
console.log("Step 3:", step3);
console.log("Step 4:", step4);
console.log("Step 5:", step5);
console.log("Final integrated result:", final_result);
console.log("Integration time:", integration_time);

// === FINAL BENCHMARK RESULTS ===
let end_time = Math.floor(Date.now() * 1000);
let total_elapsed = end_time - start_time;

console.log("=== BENCHMARK COMPLETE ===");
console.log("Total execution time:", total_elapsed);
console.log("Operations completed: 200+");
console.log("Variables used: 50+");
console.log("Arithmetic operations: 100+");
console.log("Print statements: 40+");
console.log("Timestamp operations: 20+");
console.log("Final benchmark score:", total_elapsed + final_result);
console.log("=== Orus Comprehensive Benchmark Complete ===");
